sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/WritableColumnVector.java

```diff
@@ -661,21 +671,22 @@ public final int appendArray(int length) {
    * common non-struct case.
    */
   public final int appendStruct(boolean isNull) {
+    reserve(elementsAppended + 1);
     if (isNull) {
       // This is the same as appendNull but without the assertion for struct types
-      reserve(elementsAppended + 1);
       putNull(elementsAppended);
-      elementsAppended++;
       for (WritableColumnVector c: childColumns) {
-        if (c.type instanceof StructType) {
+        if (c.isStruct()) {
           c.appendStruct(true);
         } else {
           c.appendNull();
```

Comment:
Hmm, I think the main purpose of this PR is to "avoiding allocating space in child vectors when an element is null." Here we still call `appendNull` which does `reserve` that will allocate more space if needed. So I am not sure if we avoid allocating space? Or it is in other place I missed?

Reply:
`reserve` doesn't affect child columns though, it only affects `null` and `offset` array.

Reply:
Hm? this `reserve` is called on child columns. Then it goes to `reserveInternal`. For example, in `OnHeapColumnVector`, `reserveInternal` will re-allocate on heap array which is used to store column data. Am I missing anyplace here?

Reply:
Oh I see. I think this PR doesn't change the behavior of `append` - it only changes the `put` API which is the main target and will be used by the Parquet work. I remember I wanted to minimize the impact, since `append` is used in columnar-to-row transition.

I can have a follow-up PR to change `append` too if that's useful.

Reply:
To confirm: now we have two ways to construct writable column vectors (`append` or `put`), which can result to column vectors representing same values but with different memory layout (null struct is stored differently)?

Reply:
Yes, for `append` it still add null in child vectors, for null struct.

Reply:
I can update the PR with the `append` changes too. It seems pretty straightforward.

Reply:
Yea let's update `append` as well. I think it's better if the `ColumVector` data format is normalized: we only have one way to represent the same values.